azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -351,7 +233,8 @@ def __init__(self, pipeline_configuration):
         self.pipeline_configuration = pipeline_configuration
 
     def run_op(self, op):
-        op.callback = pipeline_thread.invoke_on_callback_thread_nowait(op.callback)
+        # CT-TODO: make this more elegant
+        op.callbacks[0] = pipeline_thread.invoke_on_callback_thread_nowait(op.callbacks[0])
```

Comment:
>op.callbacks[0] [](start = 75, length = 15)

is this the top of the stack ?

Reply:
It is, however, `.pop()` removes the item from the stack, which we do not want.

I agree however that this is not ideal at all, and want to come up with something better.

---
In reply to: [346992405](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r346992405) [](ancestors = 346992405)

Reply:
hmm...this seems to be the only way to do peek....

---
In reply to: [347044769](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347044769) [](ancestors = 347044769,346992405)

Reply:
Actually, now that I think about it, we could pop and re-append... that said, I feel like that's maybe a little less clear about what's happening?

---
In reply to: [347050315](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347050315) [](ancestors = 347050315,347044769,346992405)

Reply:
my $.02: popping, adding the wrapper, and then pushing again seems completely clear to me.

---
In reply to: [347050862](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347050862) [](ancestors = 347050862,347050315,347044769,346992405)